### ISSUE TITLE
Add easier support for ObjC blocks in p/invokes

### ISSUE THREAD

**Issue by spouliot** [xamarin/xamarin-macios#6558]
It's a lot of boilerplate code that's easy to get wrong (at compile or run time).

Either 
* have sharpie generate the required code;
* have the generator (who already does it for ObjC) spit out the internal code needed based on a signature

I think I prefer the later... since the code already exists there (and any change would need to be done only once).

**Comment by rolfbjarne** [comment]
See also https://github.com/xamarin/xamarin-macios/issues/4543

**Comment by rolfbjarne** [comment]
Duplicate of #15783.
